### ISSUE TITLE
Merging raw input with non-raw layers

### ISSUE THREAD

**Issue by antonmbk** [fchollet/keras#517]
With the following code I see that it is quite easy to process two separate inputs of size (n,784) and (n,14)  to an output of size (n,17+31)=(n,48):

left = Sequential()
left.add(Dense(784, 17))
left.add(Activation('relu'))

right= Sequential()
right.add(Dense(14, 31))
right.add(Activation('relu'))

model = Sequential()
model.add(Merge([left, right], mode='concat'))

but what about if we want to just pass in raw input for one of the merged models like:

left = Sequential()
left.add(Dense(784, 17))
left.add(Activation('relu'))

right= Sequential()
(something should go here?? or not??)

model = Sequential()
model.add(Merge([left, right], mode='concat'))

Should I put a right.add(Activation('linear')) there? I was under the impression that Activation('linear') mapped the previous input to the output.  I tried it and it did not work however.  What is the way we would pass the raw input in for one of the merged Sequentials? Thoughts?

By the way we would prefer not to use Graph, because we would like to just add some normal layers to model after the Merge layer.

Thank you!

**Comment by EderSantana** [comment]
It didn't work because `Activation` does not have an `input` property. Which means it cannot be the first layer of a model. Try the following hack
```python
pass = Activation('linear')
pass.input = T.matrix()
right = Sequential()
model.add(pass)
```

Otherwise try use the following class
```python
class Pass(MaskedLayer):
    """ Do literally nothing
        Can be the first layer
    """
    def __init__(self, ndim=2):
        super(Pass, self).__init__()
        self.input = ndim_tensor(ndim)

    def get_output(self, train=False):
        X = self.get_input(train)
        return X
```

Just in case you are interested. The class above is part of [EderSantana/seya](https://github.com/EderSantana/seya/blob/master/seya/layers/base.py#L13-L23) . A companion library to Keras with some extra stuff.

**Comment by antonmbk** [comment]
I am away from my machine for a few days but this looks live a very decent workaround  and will try for my purposes the moment I am back. Thank you! Quick question : why is this capability (or something similar to a null/pass layer that has an input variable) not implemented in some way in official keras? Does it make the code  a bit slower to add an extra input linked to an output when compiling a theano function?

**Comment by EderSantana** [comment]
No, maybe its just that you and me were the first to need that. If more people look for it, we just send a pull request to the main repo. 

Do you want to do that as your first PR to keras? Talk to @fchollet about it. Bring up the idea at keras's email list.

**Comment by lemuriandezapada** [comment]
I think you could also just dump a "reshape" layer in there as an intermediate layer so you don't have to come up with new classes or perform dense computations.

**Comment by EderSantana** [comment]
`reshape` can't be used as a first layer. See here https://github.com/EderSantana/keras/blob/master/keras/layers/core.py#L286-L289

**Comment by fchollet** [comment]
Technically, you could develop a version of Reshape where you could
explicitly specify the expected input shape, hence making it usable as a
first layer.

**Comment by EderSantana** [comment]
I agree with @fchollet. It is `self.input = ndim_tensor(ndim)` all that such a new `Reshape` layer would need.

**Comment by antonmbk** [comment]
To continue on this subject, I am quite new to Keras, so this is probably a naive thing to say, but the inflexible approach for merging (can only sum and concatenate) and the fact that (to my knowledge) forking is not possible means that the library could use some API adjustments.  I myself have written some extra layers before (PRelu for convolutional layers, etc) but to write layers that depend on combining network layers or inputs is a considerable undertaking.  Does anyone share my thoughts or am I missing something here??

**Comment by EderSantana** [comment]
@antonmbk you may want to discuss that subject here #576.

**Comment by jamesmf** [comment]
Is there a good way to do this with recurrent nets?

I tried a simple 
```
flatModel = Sequential()
flatModel.add(Reshape((150,13),input_shape=(150,13)))
```
and then merging it with the other recurrent setup. here 150 is the number of time steps and 13 is the dimensionality of the raw vector. This compiles fine, but at training time complains that it expects a 2 dimensional input and was provided with 3-dim `(batch_size, 150, 13)`
